### PR TITLE
Fix #1444: disallow using nullables/undefineds as booleans (WAT)

### LINKED ISSUE #1444: card-reducer branch: morphing to red doesn't work

### PR DESCRIPTION
**Pull request by robertos**

### CODE REVIEW

File: client/src/game/ui/HanabiCard.ts [hunk 1]
```diff
@@ -687,7 +687,7 @@ export default class HanabiCard extends Konva.Group implements NodeWithTooltip {
 
   animateToDeck() {
     const layoutChild = this.parent as unknown as LayoutChild;
-    if (!layoutChild || !layoutChild.parent) {
+    if (layoutChild === undefined || !layoutChild.parent) {
```
Comment (Zamiell) [inline client/src/game/ui/HanabiCard.ts:690]: should this be
if (layoutChild === undefined || layoutChild.parent === undefined) {

Comment (Zamiell) [reply client/src/game/ui/HanabiCard.ts:690]: if (layoutChild === undefined || layoutChild === null || layoutChild.parent === undefined || layoutChild.parent === null) {

Comment (robertos) [reply client/src/game/ui/HanabiCard.ts:690]: parent can be null or undefined, i think.  I can add another condition here, but the linter allowed this one...

File: client/src/game/data/suitsInit.ts [hunk 2]
```diff
@@ -92,7 +92,7 @@ export default function suitsInit(COLORS: Map<string, Color>) {
           throw new Error(`The color "${colorString}" in the suit "${suitName}" does not exist.`);
         }
       }
-    } else if (!suitJSON.allClueColors && !suitJSON.noClueColors) {
+    } else if (suitJSON.allClueColors !== true && suitJSON.noClueColors !== true) {
```
Comment (Zamiell) [inline client/src/game/data/suitsInit.ts:95]: lol, use === false

File: client/src/game/ui/CardLayout.ts [hunk 3]
```diff
@@ -135,7 +135,7 @@ export default class CardLayout extends Konva.Group {
             opacity: 1,
             easing: Konva.Easings.EaseOut,
             onFinish: () => {
-              if (!card || !layoutChild) {
+              if (card === undefined || layoutChild === undefined) {
```
Comment (Zamiell) [inline client/src/game/ui/CardLayout.ts:138]: add both null checks

File: client/src/game/data/suitsInit.ts (hunk 2 above)
Comment (robertos) [reply client/src/game/data/suitsInit.ts:95]: NOPE: !== true => can be false or undefined. This is exactly the bug!!

File: client/src/game/ui/CardLayout.ts (hunk 3 above)
Comment (robertos) [reply client/src/game/ui/CardLayout.ts:138]: those can't be null, sorry :)

File: client/src/game/data/suitsInit.ts (hunk 2 above)
Comment (Zamiell) [reply client/src/game/data/suitsInit.ts:95]: write a comment then explaining the code that this isn't a bug

File: client/src/game/ui/CardLayout.ts (hunk 3 above)
Comment (robertos) [reply client/src/game/ui/CardLayout.ts:138]: their types are HanabiCard and Konva.Node, none of which is nullable.

File: client/src/game/ui/HanabiCard.ts [hunk 4]
```diff
@@ -1054,14 +1058,14 @@ export default class HanabiCard extends Konva.Group implements NodeWithTooltip {
         return;
       }
       const hand = this.parent.parent as unknown as CardLayout;
```
Comment (Zamiell) [inline client/src/game/ui/HanabiCard.ts:1060]: null check here, and check for this.parent before assigning this.parent.parent (?)

File: client/src/game/ui/PlayStack.ts [hunk 5]
```diff
@@ -44,7 +44,7 @@ export default class PlayStack extends Konva.Group {
         opacity,
         easing: Konva.Easings.EaseOut,
         onFinish: () => {
-          if (!layoutChild || !card || !card.parent) {
+          if (layoutChild === undefined || card === undefined || card.parent === undefined) {
```
Comment (Zamiell) [inline client/src/game/ui/PlayStack.ts:47]: null check here

File: client/src/game/ui/HanabiCard.ts (hunk 4 above)
Comment (robertos) [reply client/src/game/ui/HanabiCard.ts:1060]: no need, hand will be undefined if any of those happen

File: client/src/game/ui/PlayStack.ts (hunk 5 above)
Comment (robertos) [reply client/src/game/ui/PlayStack.ts:47]: nope, can't be null. you gotta check the types

Comment (Zamiell) [reply client/src/game/ui/PlayStack.ts:47]: card.parent can't be null?
then why do you check if it is null in other places

Comment (robertos) [reply client/src/game/ui/PlayStack.ts:47]: Because Konva, as you know, is very consistent. Parent can be null in some types, and can't in others.

Comment (Zamiell) [reply client/src/game/ui/PlayStack.ts:47]: should we just check for parent === null everywhere? it makes the code more consistent, and could make the code more resilient if we ever upgrade konva

Comment (robertos) [reply client/src/game/ui/PlayStack.ts:47]: That's the thing, it doesn't really, because TS will catch if you forget a null check.

**Comment by Zamiell** [comment]
so what was the actual bug fix amidst all of the noise?

**Comment by robertos** [comment]
> so what was the actual bug fix amidst all of the noise?

HanabiCardClick 
  // We want an exact match, so fullNote is sent as an empty string
  const cardIdentity = notes.cardIdentityFromNote(globals.variant, cardText, '');
  if (!cardIdentity.suitIndex || !cardIdentity.rank) {
  if (cardIdentity.suitIndex === null || cardIdentity.rank === null) {
    window.alert('You entered an invalid card.');
    return;
  }

**Pull request merged**